lib/internal/modules/esm/resolve.js

```diff
@@ -1,26 +1,611 @@
+        if (target !== undefined)
+          return resolveExportsTarget(packageJSONUrl, target, '', '', base);
+      }
+
+      throw new ERR_PACKAGE_PATH_NOT_EXPORTED(packageJSONUrl, '.');
+    }
+    if (packageConfig.main !== undefined) {
+      const resolved = new URL(packageConfig.main, packageJSONUrl);
+      const path = fileURLToPath(resolved);
+      if (tryStatSync(path).isFile()) return resolved;
+    }
+    if (getOptionValue('--experimental-specifier-resolution') === 'node') {
+      if (packageConfig.main !== undefined) {
+        return finalizeResolution(
+          new URL(packageConfig.main, packageJSONUrl), base);
+      } else {
+        return finalizeResolution(
+          new URL('index', packageJSONUrl), base);
+      }
+    }
+    if (packageConfig.type !== 'module') {
+      return legacyMainResolve(packageJSONUrl, packageConfig);
+    }
+  }
+
+  throw new ERR_MODULE_NOT_FOUND(
+    fileURLToPath(new URL('.', packageJSONUrl)), fileURLToPath(base));
+}
+
+
+function packageExportsResolve(
+  packageJSONUrl, packageSubpath, packageConfig, base) /* -> URL */ {
+  const exports = packageConfig.exports;
+  if (exports === undefined ||
+      isConditionalExportsMainSugar(exports, packageJSONUrl, base)) {
+    throwExportsNotFound(packageSubpath, packageJSONUrl, base);
+  }
+
+
+  if (ObjectPrototypeHasOwnProperty(exports, packageSubpath)) {
+    const target = exports[packageSubpath];
+    const resolved = resolveExportsTarget(
+      packageJSONUrl, target, '', packageSubpath, base);
+    return finalizeResolution(resolved, base);
+  }
+
+  let bestMatch = '';
+  const keys = ObjectGetOwnPropertyNames(exports);
+  for (let i = 0; i < keys.length; i++) {
+    const key = keys[i];
+    if (key[key.length - 1] !== '/') continue;
+    if (StringPrototypeStartsWith(packageSubpath, key) &&
+        key.length > bestMatch.length) {
+      bestMatch = key;
+    }
+  }
+
+  if (bestMatch) {
+    const target = exports[bestMatch];
+    const subpath = StringPrototypeSubstr(packageSubpath, bestMatch.length);
+    const resolved = resolveExportsTarget(
+      packageJSONUrl, target, subpath, packageSubpath, base);
+    return finalizeResolution(resolved, base);
+  }
+
+  throwExportsNotFound(packageSubpath, packageJSONUrl, base);
+}
+
+function getPackageType(url) {
+  const packageConfig = getPackageScopeConfig(url, url);
+  return packageConfig.type;
+}
+
+function packageResolve(specifier /* string */, base /* URL */) { /* -> URL */
+  let separatorIndex = StringPrototypeIndexOf(specifier, '/');
+  let validPackageName = true;
+  let isScoped = false;
+  if (specifier[0] === '@') {
+    isScoped = true;
+    if (separatorIndex === -1 || specifier.length === 0) {
+      validPackageName = false;
+    } else {
+      separatorIndex = StringPrototypeIndexOf(
+        specifier, '/', separatorIndex + 1);
+    }
+  }
+
+  const packageName = separatorIndex === -1 ?
+    specifier : StringPrototypeSubstr(specifier, 0, separatorIndex);
```

Comment:
any reason to use the annex b substr, instead of slice?
```suggestion
    specifier : StringPrototypeSlice(specifier, 0, separatorIndex);
```